### PR TITLE
use `ParsedLogPath`s for files in `LogSegment`

### PR DESCRIPTION
**Pull request by zachschuermann**
## What changes are proposed in this pull request?

Use `ParsedLogPath<FileMeta>`s instead of raw `FileMeta`s for `commit_files` and `checkpoint_files` fields in `LogSegment`.

Instead of
```rust
struct LogSegment {
    commit_files: Vec<FileMeta>,
    checkpoint_files: Vec<FileMeta>,
    // ...
}
```
we now have
```rust
struct LogSegment {
    commit_files: Vec<ParsedLogPath<FileMeta>>,
    checkpoint_files: Vec<ParsedLogPath<FileMeta>>,
    // ...
}
```

## How was this change tested?

exisitng

### CODE REVIEW

File: kernel/src/snapshot.rs [hunk 1]
```diff
@@ -104,14 +104,8 @@ impl Snapshot {
...
             log_root: log_url,
-            commit_files: commit_files
-                .into_iter()
-                .map(|log_path| log_path.location)
-                .collect(),
-            checkpoint_files: checkpoint_files
-                .into_iter()
-                .map(|log_path| log_path.location)
-                .collect(),
+            commit_files: commit_files.into_iter().collect(),
```
Comment (OussamaSaoudi-db) [inline kernel/src/snapshot.rs:107]: this can be shortened to: 
```
commit_files,
checkpoint_files
```

**Review by OussamaSaoudi-db — APPROVED** [review]
Besides those two comments, LGTM.

File: kernel/src/log_segment.rs [hunk 2]
```diff
@@ -15,9 +15,9 @@ use url::Url;
 pub(crate) struct LogSegment {
     pub log_root: Url,
     /// Reverse order sorted commit files in the log segment
-    pub commit_files: Vec<FileMeta>,
+    pub commit_files: Vec<ParsedLogPath<FileMeta>>,
```
[the remark this answers is not in the document]
Comment (zachschuermann) [reply kernel/src/log_segment.rs:18]: yup I originally had this way but decided to keep the type param to explicitly show it was a `ParsedLogPath` of `FileMeta` but actually probably doesn't really matter that this is explicit...

File: kernel/src/log_segment.rs [hunk 3]
```diff
@@ -43,7 +43,11 @@ impl LogSegment {
         let commit_stream = engine
             .get_json_handler()
             .read_json_files(
-                &self.commit_files,
+                &self
+                    .commit_files
+                    .iter()
+                    .map(|f| f.location.clone())
+                    .collect::<Vec<FileMeta>>(),
```
Comment (zachschuermann) [inline kernel/src/log_segment.rs:50]: optimization todo: #473

File: kernel/src/log_segment.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zachschuermann) [reply kernel/src/log_segment.rs:18]: yea actually just removed - no need for that :)

**Review by nicklan — APPROVED** [review]

File: kernel/src/log_segment.rs [hunk 4]
```diff
@@ -52,7 +56,11 @@ impl LogSegment {
         let checkpoint_stream = engine
             .get_parquet_handler()
             .read_parquet_files(
-                &self.checkpoint_files,
+                &self
+                    .checkpoint_files
+                    .iter()
+                    .map(|f| f.location.clone())
+                    .collect::<Vec<FileMeta>>(),
```
Comment (scovich) [inline kernel/src/log_segment.rs:63]: nit: cleaner to split the logic?
```rust
let checkpoint_files: Vec<_> = self
    .checkpoint_files
    .iter()
    .map(|f| f.location.clone())
    .collect();
let checkpoint_stream = engine
    .get_parquet_handler()
    .read_parquet_files(checkpoint_files, checkpoint_read_schema, meta_predicate)?
    .map_ok(...);
```

**Pull request merged**